Add json as one of possible outputs

s-and-witch opened the pull request:
Problem: we want to use diff in pipes, but default printing format isn't parse-friendly

Solution: add option "--json" and serialization of diff tree to this format.

s-and-witch committed:
Add json as one of possible outputs

Problem: we want to use diff in pipes, but default printing format isn't
parse-friendly

Solution: add option "--json" and serialization of diff tree to this format.

s-and-witch commented on src/Main.hs:
```diff
@@ -105,12 +119,13 @@ parserInfo =
 renderDiff :: RenderRunner -> RenderContext -> DerivationDiff -> IO ()
 renderDiff HumanReadable context derivation
   = Control.Monad.Reader.runReaderT (unRender (renderDiffHumanReadable derivation))  context
+renderDiff JSON _ derivation = Data.ByteString.Char8.putStrLn (Data.Aeson.encode derivation)
```
I'm not sure about output encoding. AFAIK, diff algho works with UTF-8 files

rvem commented on src/Diff/Types.hs:
```diff
@@ -0,0 +1,199 @@
+        -- ^ Will be Nothing, if environment comparison is skipped
+      }
+  deriving stock Generic
+  deriving anyclass ToJSON
```
I think it might be useful to also provide `FromJSON` instances so that `nix-diff` can also be used as a library and these diffs can be de-serialized back to Haskell types since raw JSON representation is quite hardly-readable

rvem commented:
As always, would be nice to see some tests. For JSON representation we can add golden tests as well and probably `toJSON . fromJSON` roundtrip to check that values don't change during it

s-and-witch commented:
> probably toJSON . fromJSON roundtrip to check that values don't change during it

I'm changing values during toJSON [here](https://github.com/serokell/nix-diff/blob/Player205/%2358-Add-json-output/src/Diff/Types.hs#L36) because in other case there would be a list full of `... { "type": "Both", "content": "Brethren," }, { "type": "Both", "content": " " }, { "type": "Both", "content": "or" } ...` for each line/word/character of sources (depend on Orientation value)

rvem commented:
Hmm, okay, but what is the motivation behind such serialization? I think JSON was never intended to be human-readable and it's more-or-less okay to have some boilerplate in JSON (as long as this boilerplate is not hundreds of MBs :sweat_smile:)

s-and-witch commented:
> Hmm, okay, but what is the motivation behind such serialization?

I thought that having possibility to read JSON output by eyes is better, that have no it.

s-and-witch commented:
> as long as this boilerplate is not hundreds of MBs 

Suppose, that we have source file about 1000 lines of code, that have 30-40 characters per line (some lines are empty and other a longer than 40 characters) in average, and we enable `--character-oriented` option.
In this case we would have output like `{"type":"Both","content":"a"},` for each character. There are 30 characters, so file will increase from 30-40 kb up to 900kb-1.2mb.

rvem commented on src/Diff/Types.hs:
```diff
@@ -0,0 +1,199 @@
+    where
+      -- We want to have the longest difference spans,
+      -- instead of printing each word as one item
+      squashDiff (Patience.Old a : Patience.Old b : xs) =
```
Will this squashing affect the textual (aka human-readable) representation of the given `TextDiff`?

s-and-witch replied:
I'm not sure, let's test it

rvem replied:
If it doesn't, maybe it makes sense to always do this compression when traversing through derivations while constructing diff

s-and-witch replied:
As I can see, it doesn't affect in NotTTY, and it may affect on IsTTY (colored output)

s-and-witch replied:
Oh, and it will affect on arguments output, so we should remove this squashing for them and test sources/environment

rvem replied:
This is quite weird. I'd expect that the deserialized value to act the same way as value before serialization

s-and-witch commented:
@rvem I updated PR and for now removed squashing of Diff, because after `decode . encode` test reports an error:
```
170c170
<     ←/nix/store/9dmp71si62hsgf2bi5g1ibkqdbwsv0zv-second_derivation←← ←←/nix/store/gvvxj52wd0s986ybx06gzrj8r2623fi9-third_derivation←← ←←/nix/store/5agf2p8f84fijdnlqy5mvwza3s2lic8l-old←← ←←/nix/store/10gqfacshzjzz0aayx29w5wrbcsbsa83-outputs←→/nix/store/1zy94ghyzk4f1r3alk4yhk2yvrm6x4ch-second_derivation→→ →→/nix/store/0xq8xmcxshwkai4rn55zb780pq8qdrbx-third_derivation→→ →→/nix/store/gz8x5pzvn6wpzk74m5nr7758iz8cvbsr-new→→ →→/nix/store/kqgpkfil6919d95c2vfyvdpb0fcfam30-outputs-lib→
---
>     ←/nix/store/9dmp71si62hsgf2bi5g1ibkqdbwsv0zv-second_derivation /nix/store/gvvxj52wd0s986ybx06gzrj8r2623fi9-third_derivation /nix/store/5agf2p8f84fijdnlqy5mvwza3s2lic8l-old /nix/store/10gqfacshzjzz0aayx29w5wrbcsbsa83-outputs←→/nix/store/1zy94ghyzk4f1r3alk4yhk2yvrm6x4ch-second_derivation /nix/store/0xq8xmcxshwkai4rn55zb780pq8qdrbx-third_derivation /nix/store/gz8x5pzvn6wpzk74m5nr7758iz8cvbsr-new /nix/store/kqgpkfil6919d95c2vfyvdpb0fcfam30-outputs-lib→
```
As you can see, each word and space was tagged by arrows as new/old, but after squashing these diffs are merged together into one "new" and one "old" diff.
It seems to be a good improvement, but I'm not sure, that it's ok to change current behavior.

I also want to ask you, what is the best way to test tree preservation after `decode . encode`? Now all the code is in the executable component, so, AFAIU, I can't just make it dependency of test component, so I need to create a library and move all the code into it, right?

rvem commented:
> I also want to ask you, what is the best way to test tree preservation after `decode . encode`?

Ideally, you should be able to generate some arbitrary diff structures and provide some property tests. I believe that this part is pure and thus can be extracted into library and imported in test-suite

rvem commented on golden-tests/run-test.sh:
```diff
@@ -1,9 +1,22 @@
+export OLD_DRV="$(nix-instantiate ./old-derivation/drv.nix)"
+export NEW_DRV="$(nix-instantiate ./new-derivation/drv.nix)"
+
+retval="0"
```
Oh, btw, this looks like a perfect use-case for [`bats`](https://github.com/bats-core/bats-core). What do you think about reworking this script to use it instead?

rvem commented on src/Diff/Types.hs:
```diff
@@ -0,0 +1,252 @@
+
```
Looks like a leftover

rvem commented on src/Diff/Types.hs:
```diff
@@ -0,0 +1,252 @@
+newtype TextDiff = TextDiff {unTextDiff :: [Patience.Item Text]}
+
+instance ToJSON TextDiff where
+  toJSON = listValue itemToJSON . squashDiff . unTextDiff
```
Let's leave a TODO here to decide what to do with the squashing later

rvem commented on golden-tests/helpers/run-diff.sh:
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env bash
+
+expected=$1
```
Please make sure that your script is `shellcheck` complaint

rvem commented on golden-tests/run-test.sh:
```diff
@@ -1,9 +1,12 @@
-code="$?"
-rm /tmp/nix-diff-output
-exit "$code"
+export OLD_DRV="$(nix-instantiate ./old-derivation/drv.nix)"
```
Such commands are usually extracted into `setup()`:
```
setup() {
  export OLD_DRV="$(nix-instantiate ./old-derivation/drv.nix)"
  export NEW_DRV="$(nix-instantiate ./new-derivation/drv.nix)"
}
```

rvem commented on src/Diff/Types.hs:
```diff
@@ -0,0 +1,319 @@
+{-# LANGUAGE DuplicateRecordFields #-}
```
Oh-oh, do we really need it?

rvem commented on app/Main.hs:
```diff
@@ -24,11 +24,14 @@ import qualified System.Posix.IO
 import qualified System.Posix.Terminal
```
There is an unused `import Data.Monoid ((<>))` in this module. Would be nice to catch such warnings as compile errors

rvem reviewed (approved):
LGTM, only a few nitpicks regarding testing scripts and some Haskell code cleanups. Feel free to merge PR once you deal with them.

Also, please make sure that all files have a newline at the end
